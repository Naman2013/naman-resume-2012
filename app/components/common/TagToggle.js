import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { WithContext as ReactTags } from 'react-tag-input';
import { white, pink } from '../../styles/variables/colors';

class TagToggle extends Component {

  getActiveTagIndex = (tag) => {
    return this.props.activeTags.indexOf(tag);
  }

  toggleTag = (value) => {
    const { activeTags } = this.props;
    const activeTagIndex = this.getActiveTagIndex(value);
    if (activeTagIndex > -1) {
      activeTags.splice(activeTagIndex, 1);
    } else {
      activeTags.push(value);
    }

    this.props.handleTagClick(this.props.param, activeTags);
  }

  get tagListFormatted() {
    const {
      tagList
    } = this.props;
    return tagList.map((tag, id) => ({ id, text: tag }));
  }
  render() {
    const showTags = this.tagListFormatted.length > 0;
    return (
      <div className="react-tags-container">
        {!showTags && <div>There are no tags.</div>}
        { showTags && this.tagListFormatted.map(tag => (
          <button
            key={tag.id}
            className={classnames('tag', 'btn', {
              active: this.getActiveTagIndex(tag.text) > -1
            })}
            onClick={() => { this.toggleTag(tag.text); }}
          >
            {tag.text}
          </button>)
        )}
        <style jsx>
          {`
            .react-tags-container {
              padding: 10px;
              text-transform: none;
            }
            .tag {
              display: inline-block;
              background: ${white};
              border-radius: 8px;
              box-shadow: 0px 0px 4px 0px #999;
              padding: 4px 15px 4px 10px;
              margin: 4px 8px;
              position: relative;
              outline: 0;
            }

            .tag:focus {
              outline: 0;
            }

            .active {
                border-color: ${pink};
            }
          `}
        </style>
      </div>
    );
  }
}

TagToggle.defaultProps = {
  tagList: [],
  activeTags: [],
  param: ''
};

TagToggle.propTypes = {
  activeTags: PropTypes.arrayOf(PropTypes.string),
  tagList: PropTypes.arrayOf(PropTypes.string),
  handleTagClick: PropTypes.func.isRequired,
  param: PropTypes.string
};

export default TagToggle;
