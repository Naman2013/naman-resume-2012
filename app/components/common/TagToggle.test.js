import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import TagToggle from './TagToggle';

it('it should render a menu', () => {
  const handleToggleClickSpy = sinon.spy();
  const testData = {
    handleTagClick: handleToggleClickSpy,
    activeTags: []
  };
  const tree = renderer.create(
    <TagToggle
      {...testData}
    />,
  );

  expect(tree.toJSON()).toMatchSnapshot();
});

it('it should render a defined menu', () => {
  const handleToggleClickSpy = sinon.spy();
  const testData = {
    handleTagClick: handleToggleClickSpy,
    activeTags: []
  };
  const component = shallow(<TagToggle {...testData} />);
  expect(component).toBeDefined();
});

it('it should render a no tags message if there are no tags', () => {
  const handleToggleClickSpy = sinon.spy();
  const testData = {
    handleTagClick: handleToggleClickSpy,
    tagList: [],
    activeTags: [],
  };
  const component = shallow(<TagToggle {...testData} />);
  expect(component.text()).toContain('There are no tags');
});

it('it should render tags message if tags are there', () => {
  const handleToggleClickSpy = sinon.spy();
  const testData = {
    handleTagClick: handleToggleClickSpy,
    tagList: ['yoo', 'alright'],
    activeTags: [],
  };
  const component = shallow(<TagToggle {...testData} />);
  expect(component.text()).not.toContain('There are no tags');
  expect(component.find('button').length).toBe(2);
});

it('it should call handleTagClick when a tag is clicked', () => {
  const handleToggleClickSpy = sinon.spy();
  const testData = {
    handleTagClick: handleToggleClickSpy,
    tagList: ['yoo', 'alright'],
    param: 'testParam',
    activeTags: []
  };
  const component = shallow(<TagToggle {...testData} />);
  const button = component.find('button').first();
  button.props().onClick();
  expect(handleToggleClickSpy.getCall(0).args[0]).not.toBe(testData.params, [testData.tagList[0]]);
});

it('it should have an active class if it is selected', () => {
  const testData = {
    handleTagClick: () => {},
    tagList: ['yoo', 'alright'],
    activeTags: ['yoo'],
  };
  const component = shallow(<TagToggle {...testData} />);
  const button = component.find('button').first();
  expect(button.props().className).toContain('active');
});
