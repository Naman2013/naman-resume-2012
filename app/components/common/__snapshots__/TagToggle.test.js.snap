// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render a menu 1`] = `
<div
  className="jsx-3432551512 react-tags-container"
>
  <div
    className="jsx-3432551512"
  >
    There are no tags.
  </div>
  <style
    jsx={true}
  >
    
            .react-tags-container {
              padding: 10px;
              text-transform: none;
            }
            .tag {
              display: inline-block;
              background: #fff;
              border-radius: 8px;
              box-shadow: 0px 0px 4px 0px #999;
              padding: 4px 15px 4px 10px;
              margin: 4px 8px;
              position: relative;
              outline: 0;
            }

            .tag:focus {
              outline: 0;
            }

            .active {
                border-color: #F310A7;
            }
          
  </style>
</div>
`;
